modin/experimental/batch/pipeline.py

```diff
@@ -0,0 +1,354 @@
+            The `reduce_fn` must also be specified.
+        pass_partition_id : bool, default: False
+            Whether to pass the numerical partition id to the query.
+        reduce_fn : Callable
+            The reduce function to apply if `fan_out` is set to True. This takes the `num_partition`
+            partitions that result from this query, and combines them into 1 partition.
+        output_id : int, default None
+            An id to assign to this node if it is an output.
+        """
+        self.nodes_list.append(
+            PandasQuery(
+                func,
+                is_output,
+                repartition_after,
+                fan_out,
+                pass_partition_id,
+                reduce_fn,
+            )
+        )
+        if is_output:
+            self.outputs.append(self.nodes_list[-1])
+            if output_id:
+                if self.node_to_id is None:
+                    if len(self.outputs) == 1:
+                        self.node_to_id = {}
+                    else:
+                        raise ValueError("Output ID must be specified for all nodes.")
+                self.node_to_id[self.outputs[-1]] = output_id
+            if output_id is None and self.node_to_id is not None:
+                raise ValueError("Output ID must be specified for all nodes.")
+            curr_node = self.outputs[-1]
+            curr_node.operators = self.nodes_list[:-1]
+            self.nodes_list = []
+
+    def _complete_nodes(self, list_of_nodes, ptns):
+        """
+        Run a sub-query end to end.
+
+        Parameters
+        ----------
+        list_of_nodes: list of PandasQuery
+            The functions that compose this query.
+        ptns: list of PandasOnRayDataframeVirtualPartition
+            The ptns that compose the dataframe that is input to this sub-query.
+        """
+        for node in list_of_nodes:
+            if node.fan_out:
+                if len(ptns) > 1:
+                    ErrorMessage.not_implemented(
+                        "Fan out is only supported with DataFrames with 1 partition."
+                    )
+                ptns[0] = ptns[0].force_materialization()
+                ptn_list = ptns[0].list_of_partitions_to_combine
+                ptns[0] = ptns[0].add_to_apply_calls(node.func, 0)
+                ptns[0].drain_call_queue(num_splits=1)
+                oids = []
+                new_dfs = []
+                for i in range(1, self.num_partitions):
+                    new_dfs.append(
+                        type(ptns[0])(
+                            ptn_list,
+                            full_axis=ptns[0].full_axis,
+                        ).add_to_apply_calls(node.func, i)
+                    )
+                    new_dfs[-1].drain_call_queue(num_splits=1)
+                    oids.extend(new_dfs[-1].list_of_blocks)
+                ray.wait(oids, num_returns=len(oids))
```

Comment:
Why should we wait for oids?

Reply:
We need to wait for the `oids` so that we can pass them into the `reduce` function.

Reply:
We can pass `oids` directly into the `reduce` function. Ray will take care of materializing data in the worker, where the `reduce` function is executed. But I don't see where `oids` are passed into the `reduce`. Could you point me at this?

Reply:
We just use the `new_dfs` variable above rather than passing the partitions as arguments to the reduce function below. It feels cleaner to me than passing object ids or partitions. @devin-petersohn do you think it should be changed to pass `oids` to reduce?

Reply:
By using the current approach we are blocked in the main thread because of ray.wait. By passing the partitions as arguments we are not blocked in the main thread and the partitions will be materialized in the workers.

Reply:
@RehanSD, is this addressed?

Reply:
Yeah - we no longer do any waiting. We do split the final result back up into block partitions though.